{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variables"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For convenience, there might be more than one set of samples in the cells below. If so, note that the order matters, and only the last cell will be relevant, as the previous ones will be overwritten"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "Rscript --vanilla analyze_srr.r \"SRR5090597\" \"hpv16\" \"UCSChg38\" \"../../wallaby/data/ref_genomes/hpv/HPV16.fa\" \"../../wallaby/data/ref_genomes/human/hg38.fa\" \".\" 1 5\n",
    "\n",
    "srr_name <- toString(args[1])\n",
    "donor_name <- toString(args[2])\n",
    "recipient_name <- toString(args[3])\n",
    "donor_ref_genome <- toString(args[4])\n",
    "recipient_ref_genome <- toString(args[5])\n",
    "inputs_folder <- toString(args[6])\n",
    "min_num_crossings <-args[7]\n",
    "min_num_reads <- args[8]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Variables\n",
    "srr_name <- \"SRR5090597\"\n",
    "donor_name <- \"hpv16\"\n",
    "recipient_name <- \"UCSChg38\"\n",
    "inputs_folder <- \"../../wallaby/workflows/outputs/test_sample\"\n",
    "donor_ref_genome <- \"../../wallaby/data/ref_genomes/hpv/HPV16.fa\"\n",
    "recipient_ref_genome <- \"../../wallaby/data/ref_genomes/human/hg38.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Select crossings to import and order of visualization\n",
    "crossings <- c(\"1Md_2Mr\", \"1Md_1Mr\", \"2Md_1Mr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Enable multithreading when possible (library dependent)\n",
    "options(Ncpus = parallel::detectCores())\n",
    "Sys.setenv(OMP_NUM_THREADS=toString(parallel::detectCores()))\n",
    "Sys.setenv(OMP_THREAD_LIMIT=toString(parallel::detectCores()))\n",
    "Sys.setenv(OMP_NUM_THREADS=parallel::detectCores())\n",
    "Sys.setenv(OMP_THREAD_LIMIT=parallel::detectCores())\n",
    "\n",
    "## Load or install and load all libraries\n",
    "suppressPackageStartupMessages(library(\"pacman\", character.only = TRUE))\n",
    "\n",
    "# List of CRAN packages to either Load, or Install and Load\n",
    "pacman::p_load(stringr, data.table, install = FALSE)\n",
    "\n",
    "# List of Bioconductor packages to either Load, or Install and Load\n",
    "pacman::p_load(rtracklayer, TxDb.Hsapiens.UCSC.hg38.knownGene, Repitools,\n",
    "               Organism.dplyr, install = FALSE)\n",
    "\n",
    "# Load all .bed files for all srrs created by the cromwell workflow\n",
    "load_beds <- function(srr_names, crossings, name) {\n",
    "    # Define variables to hold all srrs\n",
    "    granges_all_srrs <- list()\n",
    "\n",
    "    # Loop over each srr\n",
    "    for (srr in srr_names) {\n",
    "\n",
    "        # For each single srr\n",
    "        granges <- list()\n",
    "        granges_crossings <- list()\n",
    "\n",
    "        # Load all .bed files created by the cromwell workflow\n",
    "        for (cross in crossings){\n",
    "            # list all recipient files\n",
    "            file <- list.files(inputs_folder,\n",
    "                                      pattern=paste(srr, \n",
    "                                                    '-to-',\n",
    "                                                    name,\n",
    "                                                    \"_\", cross, \n",
    "                                                    \"_filtered.bed\", \n",
    "                                                    sep = \"\"), \n",
    "                                      recursive = TRUE, \n",
    "                                      full.names = TRUE)\n",
    "\n",
    "            # check whether the file exists and add if it does\n",
    "            if (!identical(file, character(0))) {\n",
    "                granges[[cross]] <- import(file)\n",
    "                granges_crossings[[cross]] <- cross\n",
    "            }\n",
    "\n",
    "            # add granges for this srr to the _all object\n",
    "            granges_all_srrs[[srr]] <- granges\n",
    "        }\n",
    "    }\n",
    "    \n",
    "    return (granges_all_srrs)\n",
    "}\n",
    "\n",
    "# Function to Create a Table mapping ranges of overlapping paired-end crossings \n",
    "summary_table_recipient <- function(granges, \n",
    "                                    granges_labels, \n",
    "                                    min_num_crossings,\n",
    "                                    min_num_reads, \n",
    "                                    src) {\n",
    "\n",
    "    # convert all granges to dataframes\n",
    "    granges_df <- lapply(granges, annoGR2DF)\n",
    "    # assign all granges labels (crossing names) as each dataframe's name\n",
    "    names(granges_df) <- granges_labels\n",
    "    # flatten all dataframes into a single dataframe, \n",
    "    # and use their's respective name as an identifier in a new column named 'crossing'\n",
    "    merged_df <- bind_rows(granges_df, .id = \"crossing\")\n",
    "    # convert the data.frame to a data.table\n",
    "    merged_dt <- as.data.table(merged_df)\n",
    "\n",
    "    # assign each row to a group, based on whether their ranges overlap\n",
    "    merged_dt[,group := { ir <- IRanges(start, end); subjectHits(findOverlaps(ir, reduce(ir))) }, by = chr]\n",
    "    # aggregate results by group, and add additional aggregated columns\n",
    "    merged_final <- merged_dt[, list(start = min(start), \n",
    "                                     stop = max(end), \n",
    "                                     num_crossings = length(unique(list(crossing)[[1]])),\n",
    "                                     unique_crossings = list(unique(crossing)),\n",
    "                                     num_reads = length(list(name)[[1]])\n",
    "                                     ), by = list(group, chr)]\n",
    "    \n",
    "    # filter results using a minimum number of reads and/or crossings\n",
    "    merged_final <- merged_final[merged_final[, num_reads > (min_num_reads - 1)]]\n",
    "    merged_final <- merged_final[merged_final[, num_crossings > (min_num_crossings - 1)]]\n",
    "    \n",
    "    # there might be no matches, so check for that before looking for a gene name\n",
    "    if (nrow(merged_final) != 0) {\n",
    "        # use the src database to look for the gene names or each crossing overlap region\n",
    "        # then, add it as a new column\n",
    "        merged_final$gene_name <- apply(merged_final, 1, FUN = function(x) toString(\n",
    "            unique(unlist(suppressWarnings(annoGR2DF(\n",
    "                                    transcripts(src, \n",
    "                                                 filter=~(GRangesFilter(\n",
    "                                                     GenomicRanges::GRanges(\n",
    "                                                         paste(toString(x[\"chr\"]), \":\", \n",
    "                                                               as.integer(x[\"start\"]), \"-\", \n",
    "                                                               as.integer(x[\"stop\"]), sep = \"\")))), \n",
    "                                                 columns=c(\"symbol\")))$symbol)))))\n",
    "                                        \n",
    "        # delete the 'group' column\n",
    "        merged_final <- merged_final[, !\"group\"]\n",
    "        # add an ID to each row\n",
    "        merged_final <- merged_final[, id := .I]\n",
    "        setcolorder(merged_final, c(\"id\", \"chr\", \"start\", \"stop\", \n",
    "                                    \"num_crossings\", \"unique_crossings\", \n",
    "                                    \"num_reads\", \"gene_name\"))\n",
    "                                       \n",
    "    # if there are no matches, write an NA row\n",
    "    } else {\n",
    "            return(data.table(id = \"<NA>\",\n",
    "                              chr = \"<NA>\", \n",
    "                              start = 0, \n",
    "                              stop = 0, \n",
    "                              num_crossings = 0, \n",
    "                              unique_crossings = \"<NA>\", \n",
    "                              num_reads = 0, \n",
    "                              gene_name = \"<NA>\")\n",
    "                  )\n",
    "        }\n",
    "\n",
    "    return(merged_final)\n",
    "}\n",
    "                                       \n",
    "# Function to Create a Summary Table for many SRRs\n",
    "srrs_summary_table_recipient <- function(granges_list, \n",
    "                                         min_num_crossings,\n",
    "                                         min_num_reads, \n",
    "                                         src) {\n",
    "\n",
    "    # store all tables in a single list\n",
    "    srrs_list <- list()\n",
    "    \n",
    "    # iterate over all granges\n",
    "    for (srr in names(granges_list)) {\n",
    "\n",
    "        # create a summary table for each granges object\n",
    "        granges <- unname(granges_list[srr][[1]])\n",
    "        \n",
    "        # check if there are valid files for the given SRRs\n",
    "        if (length(granges) != 0) {\n",
    "            granges_labels <- str_split(names(granges_list[srr][[1]]), \" \")\n",
    "            summary_table <- summary_table_recipient (granges = granges,\n",
    "                                                      granges_labels = granges_labels,\n",
    "                                                      min_num_crossings = 1, \n",
    "                                                      min_num_reads = min_num_reads, \n",
    "                                                      src = src)\n",
    "\n",
    "            # add a column for the srr\n",
    "            summary_table$srr <- srr\n",
    "            # add the table to the list\n",
    "            srrs_list[[srr]] <- summary_table\n",
    "        } else {\n",
    "            srrs_list[[srr]] <- data.table(srr = srr,\n",
    "                                           id = \"<NA>\",\n",
    "                                           chr = \"<NA>\", \n",
    "                                           start = 0, \n",
    "                                           stop = 0, \n",
    "                                           num_crossings = 0, \n",
    "                                           unique_crossings = \"<NA>\", \n",
    "                                           num_reads = 0, \n",
    "                                           gene_name = \"<NA>\")\n",
    "        }\n",
    "    }\n",
    "\n",
    "    # check if there are valid files for the given SRRs\n",
    "    if (length(srrs_list) != 0) {\n",
    "        # bind all the tables by row\n",
    "        srrs_summary_table <- do.call(rbind, c(srrs_list, fill=TRUE))\n",
    "        # reorder the table\n",
    "        setcolorder(srrs_summary_table, c(\"srr\", \"id\", \"chr\", \"start\", \"stop\", \n",
    "                                          \"num_crossings\", \"unique_crossings\", \n",
    "                                          \"num_reads\", \"gene_name\"))\n",
    "\n",
    "        return (srrs_summary_table)\n",
    "        } else {\n",
    "            print(\"No files matching SRRs\")\n",
    "            return\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load all .bed files for all srrs created by the cromwell workflow\n",
    "donor_granges_all_srrs <- load_beds(srr_name, crossings, donor_name)\n",
    "recip_granges_all_srrs <- load_beds(srr_name, crossings, recipient_name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Create a sqlite database from TxDb and corresponding Org packages\n",
    "# The database provides a convenient way to map between gene, transcript, and protein identifiers.\n",
    "src <- suppressMessages(suppressWarnings(src_organism(\"TxDb.Hsapiens.UCSC.hg38.knownGene\")))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Aggregated view of all overlapping crossings for the potential recipient for all srrs\n",
    "srrs_summary_table <- srrs_summary_table_recipient (granges_list = recip_granges_all_srrs, \n",
    "                                                    min_num_crossings = 1,\n",
    "                                                    min_num_reads = 5, \n",
    "                                                    src = src)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.table: 1 × 9</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>srr</th><th scope=col>id</th><th scope=col>chr</th><th scope=col>start</th><th scope=col>stop</th><th scope=col>num_crossings</th><th scope=col>unique_crossings</th><th scope=col>num_reads</th><th scope=col>gene_name</th></tr>\n",
       "\t<tr><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>SRR5090597</td><td>&lt;NA&gt;</td><td>&lt;NA&gt;</td><td>0</td><td>0</td><td>0</td><td>&lt;NA&gt;</td><td>0</td><td>&lt;NA&gt;</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.table: 1 × 9\n",
       "\\begin{tabular}{lllllllll}\n",
       " srr & id & chr & start & stop & num\\_crossings & unique\\_crossings & num\\_reads & gene\\_name\\\\\n",
       " <chr> & <chr> & <chr> & <dbl> & <dbl> & <dbl> & <chr> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t SRR5090597 & <NA> & <NA> & 0 & 0 & 0 & <NA> & 0 & <NA>\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.table: 1 × 9\n",
       "\n",
       "| srr &lt;chr&gt; | id &lt;chr&gt; | chr &lt;chr&gt; | start &lt;dbl&gt; | stop &lt;dbl&gt; | num_crossings &lt;dbl&gt; | unique_crossings &lt;chr&gt; | num_reads &lt;dbl&gt; | gene_name &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|---|---|\n",
       "| SRR5090597 | &lt;NA&gt; | &lt;NA&gt; | 0 | 0 | 0 | &lt;NA&gt; | 0 | &lt;NA&gt; |\n",
       "\n"
      ],
      "text/plain": [
       "  srr        id   chr  start stop num_crossings unique_crossings num_reads\n",
       "1 SRR5090597 <NA> <NA> 0     0    0             <NA>             0        \n",
       "  gene_name\n",
       "1 <NA>     "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "srrs_summary_table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "fwrite(srrs_summary_table, \"test_table.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
