{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "srr_names <- c(\"SRR12091998\", \"SRR12091997\", \"SRR12091996\", \"SRR12091995\", \"SRR12091994\", \"SRR12091993\", \"SRR12091992\", \"SRR12091991\", \n",
    "\"SRR12091990\")\n",
    "donor_name <- \"sars-cov-2\"\n",
    "recipient_name <- \"UCSChg38\"\n",
    "inputs_folder <- \"../../wallaby/workflows/datasets/sars-cov-2/\"\n",
    "donor_ref_genome <- \"../../wallaby/data/ref_genomes/sars-cov-2/wuhan-hu-1.fa\"\n",
    "recipient_ref_genome <- \"../../wallaby/data/ref_genomes/human/hg38.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "srr_names <- c(\"SRR11772358\", \"SRR11772359\", \"SRR11772360\", \"SRR11772361\", \"SRR11772362\", \"SRR11772363\", \"SRR11772364\", \"SRR11772365\", \n",
    "\"SRR11772366\", \"SRR11772367\", \"SRR11772368\", \"SRR11772369\", \"SRR11772370\", \"SRR11772371\", \"SRR11772372\", \"SRR11772373\", \n",
    "\"SRR11772374\", \"SRR11772375\", \"SRR11772376\", \"SRR11772377\", \"SRR11772378\", \"SRR11772379\", \"SRR11772380\", \"SRR11772381\", \n",
    "\"SRR11772382\", \"SRR11772383\", \"SRR11772384\", \"SRR11772385\", \"SRR11772386\", \"SRR11772387\", \"SRR11772388\", \"SRR11772389\", \n",
    "\"SRR11772390\", \"SRR11772391\", \"SRR11772392\", \"SRR11772393\", \"SRR11772394\", \"SRR12340057\", \"SRR12340058\", \"SRR12340059\", \n",
    "\"SRR12340060\", \"SRR12340061\", \"SRR12340062\", \"SRR12340063\", \"SRR12340064\", \"SRR12340065\", \"SRR12340066\", \"SRR12340067\", \n",
    "\"SRR12340068\", \"SRR12340069\", \"SRR12340070\", \"SRR12340071\", \"SRR12340072\", \"SRR12340073\", \"SRR12340074\", \"SRR12340075\", \n",
    "\"SRR12340076\", \"SRR12340077\", \"SRR12340078\", \"SRR12340079\", \"SRR12340080\", \"SRR12340081\", \"SRR12340082\", \"SRR12340083\", \n",
    "\"SRR12340084\", \"SRR12340085\", \"SRR12340086\", \"SRR12340087\", \"SRR12340088\", \"SRR12340089\", \"SRR12340090\", \"SRR12340091\", \n",
    "\"SRR12340092\", \"SRR12340093\", \"SRR12340094\", \"SRR12340095\", \"SRR12340096\", \"SRR12340097\", \"SRR12340098\", \"SRR12340099\", \n",
    "\"SRR12340100\", \"SRR12340101\", \"SRR12340102\", \"SRR12340103\", \"SRR12340104\", \"SRR12340105\", \"SRR12340106\", \"SRR12340107\")\n",
    "donor_name <- \"sars-cov-2\"\n",
    "recipient_name <- \"UCSChg38\"\n",
    "inputs_folder <- \"../../wallaby/workflows/datasets/sars-cov-2-biopsies/\"\n",
    "donor_ref_genome <- \"../../wallaby/data/ref_genomes/sars-cov-2/wuhan-hu-1.fa\"\n",
    "recipient_ref_genome <- \"../../wallaby/data/ref_genomes/human/hg38.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [],
   "source": [
    "## Variables\n",
    "srr_names <- \"SRR5090597,SRR5090599\"\n",
    "donor_name <- \"hpv16\"\n",
    "recipient_name <- \"UCSChg38\"\n",
    "inputs_folder <- \"../../wallaby/workflows/cromwell-grouped-inputs/\"\n",
    "donor_ref_genome <- \"../../wallaby/data/ref_genomes/hpv/HPV16.fa\"\n",
    "recipient_ref_genome <- \"../../wallaby/data/ref_genomes/human/hg38.fa\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Tertiary analysis of a Donor-to-Recipient pipeline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Setup Environment\n",
    "# Enable multithreading when possible (library dependent)\n",
    "options(Ncpus = parallel::detectCores())\n",
    "Sys.setenv(OMP_NUM_THREADS=toString(parallel::detectCores()))\n",
    "Sys.setenv(OMP_THREAD_LIMIT=toString(parallel::detectCores()))\n",
    "Sys.setenv(OMP_NUM_THREADS=parallel::detectCores())\n",
    "Sys.setenv(OMP_THREAD_LIMIT=parallel::detectCores())\n",
    "\n",
    "## Load or install and load all libraries\n",
    "suppressPackageStartupMessages(library(\"pacman\", character.only = TRUE))\n",
    "suppressPackageStartupMessages(library(\"BSgenome.Hsapiens.UCSC.hg38\", character.only = TRUE))\n",
    "\n",
    "# List of CRAN packages to either Load, or Install and Load\n",
    "pacman::p_load(dplyr, ggplot2, shiny, shinyLP, DT,  ggrepel,  tidyr, data.table, kableExtra, knitr, IRdisplay, install = FALSE)\n",
    "\n",
    "# List of Bioconductor packages to either Load, or Install and Load\n",
    "pacman::p_load(BSgenome, BSgenome.Hsapiens.UCSC.hg38, GenomicFeatures, GenomicAlignments,  Rsubread,  Rsamtools, bamsignals,  \n",
    "               rtracklayer, GenomicRanges, org.Hs.eg.db, Organism.dplyr,\n",
    "               TxDb.Hsapiens.UCSC.hg38.knownGene, regioneR, karyoploteR, seqinr, Repitools, Gviz, Biostrings, install = FALSE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in rm(srrs_donor):\n",
      "\"object 'srrs_donor' not found\"\n",
      "Warning message in rm(srrs_recipient):\n",
      "\"object 'srrs_recipient' not found\"\n"
     ]
    }
   ],
   "source": [
    "## Loading pipeline output files\n",
    "# Try to open all files - note that some will not exists as that specific crossing bucket did not return any matches\n",
    "srr_names <- strsplit(srr_names, \",\")[[1]]\n",
    "\n",
    "try(rm(srrs_donor))\n",
    "try(rm(srrs_recipient))\n",
    "\n",
    "srrs_donor <- list()\n",
    "srrs_recipient <- list()\n",
    "\n",
    "for (srr_name in srr_names){\n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', donor_name, sep = \"\"),\"_MMd_MUr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {donor_MMd_MUr <- import(file)}\n",
    "\n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', donor_name, sep = \"\"),\"_MMd_UMr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {donor_MMd_UMr <- import(file)}\n",
    "\n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', donor_name, sep = \"\"),\"_MUd_MMr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {donor_MUd_MMr <- import(file)}\n",
    "\n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', donor_name, sep = \"\"),\"_MUd_UMr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {donor_MUd_UMr <- import(file)}\n",
    "\n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', donor_name, sep = \"\"),\"_UMd_MMr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {donor_UMd_MMr <- import(file)}\n",
    "\n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', donor_name, sep = \"\"),\"_UMd_MUr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {donor_UMd_MUr <- import(file)}\n",
    "    \n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', recipient_name, sep = \"\"),\"_MMd_MUr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {recip_MMd_MUr <- import(file)}\n",
    "\n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', recipient_name, sep = \"\"),\"_MMd_UMr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {recip_MMd_UMr <- import(file)}\n",
    "\n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', recipient_name, sep = \"\"),\"_MUd_MMr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {recip_MUd_MMr <- import(file)}\n",
    "\n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', recipient_name, sep = \"\"),\"_MUd_UMr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {recip_MUd_UMr <- import(file)}\n",
    "\n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', recipient_name, sep = \"\"),\"_UMd_MMr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {recip_UMd_MMr <- import(file)}\n",
    "\n",
    "    file <- paste(inputs_folder, paste(srr_name, '-to-', recipient_name, sep = \"\"),\"_UMd_MUr.bed\", sep = \"\")\n",
    "    if (file.exists(file)) {recip_UMd_MUr <- import(file)}\n",
    "\n",
    "    donor_beds <- list()\n",
    "    donor_names <- list(\"MMd_MUr\", \"MMd_UMr\", \"MUd_MMr\", \"MUd_UMr\", \"UMd_MMr\", \"UMd_MUr\")\n",
    "\n",
    "    for (name in donor_names) {\n",
    "        object_name <- paste(\"donor_\", name, sep=\"\")\n",
    "        try(donor_beds[[name]] <- get(object_name))\n",
    "    }\n",
    "    srrs_donor[srr_name] <- list(donor_beds)\n",
    "    \n",
    "    recip_beds <- list()\n",
    "    recip_names <- list(\"MMd_MUr\", \"MMd_UMr\", \"MUd_MMr\", \"MUd_UMr\", \"UMd_MMr\", \"UMd_MUr\")\n",
    "\n",
    "    for (name in recip_names) {\n",
    "        object_name <- paste(\"recip_\", name, sep=\"\")\n",
    "        try(recip_beds[[name]] <- get(object_name))\n",
    "    }\n",
    "    srrs_recipient[srr_name] <- list(recip_beds)\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "tags": [
     "parameters"
    ]
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "\"Ignoring remaining part of query: CREATE INDEX IF NOT EXISTS entrez_accession ON id_accession (entrez);\"\n",
      "Warning message:\n",
      "\"Ignoring remaining part of query: CREATE INDEX IF NOT EXISTS entrez_protein ON id_protein (entrez);\"\n"
     ]
    }
   ],
   "source": [
    "# Helper Functions\n",
    "src <- suppressMessages(src_organism(\"TxDb.Hsapiens.UCSC.hg38.knownGene\"))\n",
    "\n",
    "summary_table <- function(table_type, srr = \"number\", granges, granges_labels, min_num_crossings = 3, min_num_reads = 3){\n",
    "    granges_df <- lapply(granges, annoGR2DF)\n",
    "    names(granges_df) <- granges_labels\n",
    "    merged_df <- bind_rows(granges_df, .id = \"crossing\")\n",
    "    merged_dt <- as.data.table(merged_df)\n",
    "\n",
    "    ## Interval per chromosome\n",
    "    merged_dt[,group := { ir <- IRanges(start, end); subjectHits(findOverlaps(ir, reduce(ir))) }, by = chr]\n",
    "    merged_final <- merged_dt[, list(start=min(start), \n",
    "                                     stop=max(end), \n",
    "                                     num_crossings=length(unique(list(crossing)[[1]])),\n",
    "                                     unique_crossings=list(unique(crossing)),\n",
    "                                     num_reads=length(list(name)[[1]])\n",
    "                                     ), by=list(group,chr)]\n",
    "    merged_final <- merged_final[merged_final[, num_reads > (min_num_reads - 1)]]\n",
    "    merged_final <- merged_final[merged_final[, num_crossings > (min_num_crossings - 1)]]\n",
    "    \n",
    "    if (table_type == \"recipient\") {\n",
    "        txdb <- TxDb.Hsapiens.UCSC.hg38.knownGene\n",
    "        Hsapiens <- BSgenome.Hsapiens.UCSC.hg38\n",
    "\n",
    "        if (nrow(merged_final) != 0){\n",
    "            merged_final$gene_name <- apply(merged_final, 1, FUN = function(x) toString(\n",
    "                unique(unlist(suppressWarnings(annoGR2DF(\n",
    "                                        transcripts(src, \n",
    "                                                     filter=~(GRangesFilter(\n",
    "                                                         GenomicRanges::GRanges(\n",
    "                                                             paste(toString(x[\"chr\"]), \":\", \n",
    "                                                                   as.integer(x[\"start\"]), \"-\", \n",
    "                                                                   as.integer(x[\"stop\"]), sep = \"\")))), \n",
    "                                                     columns=c(\"symbol\")))$symbol)))))\n",
    "\n",
    "            merged_final$srr <- srr\n",
    "\n",
    "            merged_final$sequence <- apply(merged_final, 1, FUN = function(x) toString(getSeq(Hsapiens, \n",
    "                                                                               toString(x[\"chr\"]), \n",
    "                                                                               start = as.integer(x[\"start\"]), \n",
    "                                                                               end = as.integer(x[\"stop\"]))))\n",
    "\n",
    "            merged_final <- merged_final[, !\"group\"]\n",
    "            merged_final <- merged_final[with(merged_final, order(num_reads, decreasing = TRUE)), ]\n",
    "\n",
    "            return(merged_final)\n",
    "        } else {\n",
    "            return(data.table(chr=\"<NA>\", start=0, stop=0, num_crossings=0, unique_crossings=\"<NA>\", num_reads=0, gene_name=\"<NA>\", srr=srr, sequence=\"<NA>\"))\n",
    "        }\n",
    "    }                                   \n",
    "    else if (table_type == \"donor\") {\n",
    "        if (nrow(merged_final) != 0) {\n",
    "            merged_final$srr <- srr\n",
    "            merged_final <- merged_final[, !\"group\"]\n",
    "            merged_final <- merged_final[with(merged_final, order(num_reads, decreasing = TRUE)), ]\n",
    "            return(merged_final)\n",
    "        } else {\n",
    "            return(data.table(chr=\"<NA>\", start=0, stop=0, num_crossings=0, unique_crossings=\"<NA>\", num_reads=0, srr=srr))\n",
    "        }\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "srrs_list <- list()\n",
    "\n",
    "for (srr_name in srr_names){\n",
    "    srr_table <- summary_table(table_type = \"donor\", srr = srr_name, granges = srrs_donor[srr_name][[1]], names(srrs_donor[srr_name][[1]]), min_num_crossings = 4, min_num_reads = 10)\n",
    "    srr_table$srr_name <- srr_name\n",
    "    srrs_list[[srr_name]] <- srr_table\n",
    "}\n",
    "\n",
    "srrs_table = do.call(rbind, c(srrs_list, fill=TRUE))\n",
    "srrs_table <- srrs_table[, !\"srr_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"table table-striped lightable-paper\" style='width: auto !important;  font-family: \"Arial Narrow\", arial, helvetica, sans-serif; width: auto !important; margin-left: auto; margin-right: auto;'>\n",
       " <thead>\n",
       "  <tr>\n",
       "   <th style=\"text-align:left;\"> chr </th>\n",
       "   <th style=\"text-align:right;\"> start </th>\n",
       "   <th style=\"text-align:right;\"> stop </th>\n",
       "   <th style=\"text-align:right;\"> num_crossings </th>\n",
       "   <th style=\"text-align:left;\"> unique_crossings </th>\n",
       "   <th style=\"text-align:right;\"> num_reads </th>\n",
       "   <th style=\"text-align:left;\"> srr </th>\n",
       "  </tr>\n",
       " </thead>\n",
       "<tbody>\n",
       "  <tr>\n",
       "   <td style=\"text-align:left;\"> NC_001526.4 </td>\n",
       "   <td style=\"text-align:right;\"> 7377 </td>\n",
       "   <td style=\"text-align:right;\"> 7906 </td>\n",
       "   <td style=\"text-align:right;\"> 6 </td>\n",
       "   <td style=\"text-align:left;\"> MMd_MUr, MMd_UMr, MUd_MMr, MUd_UMr, UMd_MMr, UMd_MUr </td>\n",
       "   <td style=\"text-align:right;\"> 650 </td>\n",
       "   <td style=\"text-align:left;width: 30em; max-width: 30em; \"> SRR5090597 </td>\n",
       "  </tr>\n",
       "  <tr>\n",
       "   <td style=\"text-align:left;\"> NC_001526.4 </td>\n",
       "   <td style=\"text-align:right;\"> 7143 </td>\n",
       "   <td style=\"text-align:right;\"> 7371 </td>\n",
       "   <td style=\"text-align:right;\"> 6 </td>\n",
       "   <td style=\"text-align:left;\"> MMd_MUr, MMd_UMr, MUd_MMr, MUd_UMr, UMd_MMr, UMd_MUr </td>\n",
       "   <td style=\"text-align:right;\"> 177 </td>\n",
       "   <td style=\"text-align:left;width: 30em; max-width: 30em; \"> SRR5090597 </td>\n",
       "  </tr>\n",
       "  <tr>\n",
       "   <td style=\"text-align:left;\"> NC_001526.4 </td>\n",
       "   <td style=\"text-align:right;\"> 7449 </td>\n",
       "   <td style=\"text-align:right;\"> 7906 </td>\n",
       "   <td style=\"text-align:right;\"> 6 </td>\n",
       "   <td style=\"text-align:left;\"> MMd_MUr, MMd_UMr, MUd_MMr, MUd_UMr, UMd_MMr, UMd_MUr </td>\n",
       "   <td style=\"text-align:right;\"> 94 </td>\n",
       "   <td style=\"text-align:left;width: 30em; max-width: 30em; \"> SRR5090599 </td>\n",
       "  </tr>\n",
       "  <tr>\n",
       "   <td style=\"text-align:left;\"> NC_001526.4 </td>\n",
       "   <td style=\"text-align:right;\"> 7148 </td>\n",
       "   <td style=\"text-align:right;\"> 7271 </td>\n",
       "   <td style=\"text-align:right;\"> 5 </td>\n",
       "   <td style=\"text-align:left;\"> MMd_MUr, MUd_MMr, MUd_UMr, UMd_MMr, UMd_MUr </td>\n",
       "   <td style=\"text-align:right;\"> 20 </td>\n",
       "   <td style=\"text-align:left;width: 30em; max-width: 30em; \"> SRR5090599 </td>\n",
       "  </tr>\n",
       "</tbody>\n",
       "</table>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "srrs_table %>%\n",
    "kable(\"html\") %>%\n",
    "    kable_styling(bootstrap_options = \"striped\", full_width = F, position = \"left\") %>%\n",
    "    kable_paper(full_width = F) %>%\n",
    "    column_spec(7, width = \"30em\", width_max = \"30em\") %>%\n",
    "    as.character() %>%\n",
    "    display_html()     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "srrs_list <- list()\n",
    "\n",
    "for (srr_name in srr_names){\n",
    "    srr_table <- summary_table(table_type = \"recipient\", srr = srr_name, granges = srrs_recipient[srr_name][[1]], names(srrs_recipient[srr_name][[1]]), \n",
    "                               min_num_crossings = 3, min_num_reads = 1)\n",
    "    srr_table$srr_name <- srr_name\n",
    "    srrs_list[[srr_name]] <- srr_table\n",
    "}\n",
    "\n",
    "srrs_table = do.call(rbind, c(srrs_list, fill=TRUE))\n",
    "srrs_table <- srrs_table[, !\"srr_name\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"table table-striped lightable-paper\" style='width: auto !important;  font-family: \"Arial Narrow\", arial, helvetica, sans-serif; width: auto !important; margin-left: auto; margin-right: auto;'>\n",
       " <thead>\n",
       "  <tr>\n",
       "   <th style=\"text-align:left;\"> chr </th>\n",
       "   <th style=\"text-align:right;\"> start </th>\n",
       "   <th style=\"text-align:right;\"> stop </th>\n",
       "   <th style=\"text-align:right;\"> num_crossings </th>\n",
       "   <th style=\"text-align:left;\"> unique_crossings </th>\n",
       "   <th style=\"text-align:right;\"> num_reads </th>\n",
       "   <th style=\"text-align:left;\"> gene_name </th>\n",
       "   <th style=\"text-align:left;\"> srr </th>\n",
       "   <th style=\"text-align:left;\"> sequence </th>\n",
       "  </tr>\n",
       " </thead>\n",
       "<tbody>\n",
       "  <tr>\n",
       "   <td style=\"text-align:left;\"> chr1 </td>\n",
       "   <td style=\"text-align:right;\"> 150578243 </td>\n",
       "   <td style=\"text-align:right;\"> 150578495 </td>\n",
       "   <td style=\"text-align:right;\"> 6 </td>\n",
       "   <td style=\"text-align:left;\"> MMd_MUr, MMd_UMr, MUd_MMr, MUd_UMr, UMd_MMr, UMd_MUr </td>\n",
       "   <td style=\"text-align:right;\"> 528 </td>\n",
       "   <td style=\"text-align:left;\"> MCL1 </td>\n",
       "   <td style=\"text-align:left;\"> SRR5090597 </td>\n",
       "   <td style=\"text-align:left;width: 30em; max-width: 30em; \"> CCCAGCCTCTTTGTTTAACTAGCCAGTCCCGTTTTGTCCTTACGAGAACGTCTGTGATACTTTCTGCTAATGGTTCGATGCAGCTTTCTTGGTTTATGGTCTTCAAGTGTTTAGCCACAAAGGCACCAAAAGAAATGAGAGTCACAATCCTGCCCCAGTTTGTTACGCCGTCGCTGAAAACATGGATCATCACTCGAGACAACGATTTCACATCGTCTTCGTTTTTGATGTCCAGTTTCCGAAGCATGCCTGA </td>\n",
       "  </tr>\n",
       "  <tr>\n",
       "   <td style=\"text-align:left;\"> chr1 </td>\n",
       "   <td style=\"text-align:right;\"> 150576896 </td>\n",
       "   <td style=\"text-align:right;\"> 150577494 </td>\n",
       "   <td style=\"text-align:right;\"> 6 </td>\n",
       "   <td style=\"text-align:left;\"> MMd_MUr, MMd_UMr, MUd_MMr, MUd_UMr, UMd_MMr, UMd_MUr </td>\n",
       "   <td style=\"text-align:right;\"> 386 </td>\n",
       "   <td style=\"text-align:left;\"> MCL1 </td>\n",
       "   <td style=\"text-align:left;\"> SRR5090597 </td>\n",
       "   <td style=\"text-align:left;width: 30em; max-width: 30em; \"> GCTTCCCTTGTACAGTACTGAGGCTTACAGTCATAGTTCTATTACTTGTAACTTTTACACAGGTCACTGGCATTCTTAGTGCTTCTCTTAACACTACAGTAAATTAATGAATTCGGCGGGTAATCAATTCTATGACTTGCCTGGCTACTGGCCACTTTCCTGTTCTCAACAAGGAAATTAAGTCTTTCCACCCTACCATCTTCACTAAATCTAAAAGTCCTCCTCCATAGCTTCCCAAACAAAGTTTGTTTGTTGCTGAAACTGAACTTTGCTTCTTTCAGACAGTGACTCTTCAATCAATGGGGAGCACTCTTCCCATGTATTTATTCTTGTTAGCCATAATCCTCTTGCCACTTGCTTTTCTGGCTAGGTTGCTAGGGTGCAACTCTAGGAAGTTACAGCTTGGAGTCCAACTGCATAAACTGGTTTTGGTGGTGGTGGTGGTTGGTTAAAAGTCAACTATTGCACTTACAGTAAGGCTATCTTATTAGATATGCCAAACCAGCTCCTACTCCAGCAACACCTGCAAAAGCCAGCAGCACATTCCTGATGCCACCTTCTAGGTCCTCTACATGGAAGAACTCCACAAACCCATCCTA </td>\n",
       "  </tr>\n",
       "  <tr>\n",
       "   <td style=\"text-align:left;\"> chr8 </td>\n",
       "   <td style=\"text-align:right;\"> 127828232 </td>\n",
       "   <td style=\"text-align:right;\"> 127828354 </td>\n",
       "   <td style=\"text-align:right;\"> 5 </td>\n",
       "   <td style=\"text-align:left;\"> MMd_MUr, MMd_UMr, MUd_MMr, UMd_MMr, UMd_MUr </td>\n",
       "   <td style=\"text-align:right;\"> 6 </td>\n",
       "   <td style=\"text-align:left;\"> PVT1 </td>\n",
       "   <td style=\"text-align:left;\"> SRR5090597 </td>\n",
       "   <td style=\"text-align:left;width: 30em; max-width: 30em; \"> GCTTCTTACTGAGTGTCTAGTTCCTTGTTGTACAGGTGAGGGAAACTGAGGCCCAGAGTGGATGTGACGTTTCCGAAGGAGAGGCTCTGTAAAGTTCTGAGTAGGTGAGACTTATTATTAGCT </td>\n",
       "  </tr>\n",
       "  <tr>\n",
       "   <td style=\"text-align:left;\"> chr8 </td>\n",
       "   <td style=\"text-align:right;\"> 127828082 </td>\n",
       "   <td style=\"text-align:right;\"> 127828207 </td>\n",
       "   <td style=\"text-align:right;\"> 3 </td>\n",
       "   <td style=\"text-align:left;\"> MUd_UMr, UMd_MMr, UMd_MUr </td>\n",
       "   <td style=\"text-align:right;\"> 4 </td>\n",
       "   <td style=\"text-align:left;\"> PVT1 </td>\n",
       "   <td style=\"text-align:left;\"> SRR5090597 </td>\n",
       "   <td style=\"text-align:left;width: 30em; max-width: 30em; \"> GGCAGGGTGTGTATCTTCTTGCTGCTGGGATGTTCTACTTTGTTGGACAGAAGGCTATGTGCTTTCTGGCAGGTTACACCTGGGTACATTTCAGTCAAGCCGAGCTGGCTCTCTCCTCATTTCATT </td>\n",
       "  </tr>\n",
       "  <tr>\n",
       "   <td style=\"text-align:left;\"> chr1 </td>\n",
       "   <td style=\"text-align:right;\"> 206949933 </td>\n",
       "   <td style=\"text-align:right;\"> 206950244 </td>\n",
       "   <td style=\"text-align:right;\"> 6 </td>\n",
       "   <td style=\"text-align:left;\"> MMd_MUr, MMd_UMr, MUd_MMr, MUd_UMr, UMd_MMr, UMd_MUr </td>\n",
       "   <td style=\"text-align:right;\"> 54 </td>\n",
       "   <td style=\"text-align:left;\">  </td>\n",
       "   <td style=\"text-align:left;\"> SRR5090599 </td>\n",
       "   <td style=\"text-align:left;width: 30em; max-width: 30em; \"> GGCCACCCTCCAGGCTCTAGACGATAATCCGTTGCTTGCCTCTTCTAGCTTCTGGTTGCTGCTGACATTCTTTGGTGTTCTTTGGCACGTAGCTGTGTCACTCCAACCTCTGCCTTCATAGTCACATTGGCTTCTGTGTATGTCTCAAATAACTCTCTGTTTCTCTGTTATATGGATGGTATTTAGGGCCCACCCAGCTAATCCAAGATGATGTCCTCATCTCAAGATCCTGAACTTCATCACATCTACAAAGACTCTTTTCCAAATAAGGTCACACTTCTGGGCTCCAGAGATTGAGACATAGACAAACCT </td>\n",
       "  </tr>\n",
       "  <tr>\n",
       "   <td style=\"text-align:left;\"> chr1 </td>\n",
       "   <td style=\"text-align:right;\"> 206940489 </td>\n",
       "   <td style=\"text-align:right;\"> 206940589 </td>\n",
       "   <td style=\"text-align:right;\"> 6 </td>\n",
       "   <td style=\"text-align:left;\"> MMd_MUr, MMd_UMr, MUd_MMr, MUd_UMr, UMd_MMr, UMd_MUr </td>\n",
       "   <td style=\"text-align:right;\"> 36 </td>\n",
       "   <td style=\"text-align:left;\"> PIGR </td>\n",
       "   <td style=\"text-align:left;\"> SRR5090599 </td>\n",
       "   <td style=\"text-align:left;width: 30em; max-width: 30em; \"> CTGGGAAGACCGCCAGCAGGCAGGTGAGCACGAAGAGCAGCATTGCTGGTGGGTCCCGAGCGCCGCACCACTCAGGCCGACTTCTCCTGTGCAATGCTGAA </td>\n",
       "  </tr>\n",
       "  <tr>\n",
       "   <td style=\"text-align:left;\"> chr1 </td>\n",
       "   <td style=\"text-align:right;\"> 206939232 </td>\n",
       "   <td style=\"text-align:right;\"> 206939466 </td>\n",
       "   <td style=\"text-align:right;\"> 4 </td>\n",
       "   <td style=\"text-align:left;\"> MUd_MMr, MUd_UMr, UMd_MMr, UMd_MUr </td>\n",
       "   <td style=\"text-align:right;\"> 22 </td>\n",
       "   <td style=\"text-align:left;\"> PIGR </td>\n",
       "   <td style=\"text-align:left;\"> SRR5090599 </td>\n",
       "   <td style=\"text-align:left;width: 30em; max-width: 30em; \"> GTGCCGTTCTCCGGGAAGTTGGTGAGGTTAGCCCTGCCTGCATATTTGCTGGAGACGTAGCCCTCCGAGGAGATGAGGGTTATGCAGCCACCTCTAGCTCCCTGCCGGCACCAGTACTTCCGGGTGTGCCGGTTGACAGAGGTGGGTGGGTAGTAGCACGTGATGGACACTGAGTTACCTTCCACACTATTCACCTCCTCGGGACCAAATATGGGACTCTTCGTGGAGATGGCTG </td>\n",
       "  </tr>\n",
       "</tbody>\n",
       "</table>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "srrs_table %>%\n",
    "kable(\"html\") %>%\n",
    "    kable_styling(bootstrap_options = \"striped\", full_width = F, position = \"left\") %>%\n",
    "    kable_paper(full_width = F) %>%\n",
    "    column_spec(9, width = \"30em\", width_max = \"30em\") %>%\n",
    "    as.character() %>%\n",
    "    display_html()     "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
